File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	ch := make(chan int)

	go func(t *TreeNode) {
		inorder(t, ch)
		close(ch)
	}(root)

	v := <-ch
	for w := range ch {
		if v >= w {
			return false
		}
		v = w
	}
	return true
}

func inorder(t *TreeNode, ch chan<- int) {
	if t == nil {
		return
	}
	inorder(t.Left, ch)
	ch <- t.Val
	inorder(t.Right, ch)
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Boolean
     */
    function isValidBST($root) {
        return $this->isBst($root, PHP_INT_MIN, PHP_INT_MAX);
    }
    
    function isBst ($root, $min, $max){
        if($root == null){
            return true;
        }
        
        if($root->val < $min || $root->val > $max){
            return false;
        }
        
        return $this->isBst($root->left, $min, $root->val - 1) && $this->isBst($root->right, $root->val + 1, $max);
    }
}